src/openms/source/MATH/STATISTICS/QuadraticRegression.cpp

```diff
@@ -34,21 +34,120 @@
+      const double& y = *y_begin;
+      const double& weight = *w_begin;
+      chi_squared += weight * std::pow(y - a - b * x - c * x * x, 2);
+    }
 
-    QuadraticRegression::QuadraticRegression() :
-      a_(0), b_(0), c_(0), chi_squared_(0) {}
+    return chi_squared;
+  }
 
-    double QuadraticRegression::eval(double x) const {return a_ + b_*x + c_*x*x;}
 
-    double QuadraticRegression::eval(double A, double B, double C, double x) {return A + B*x + C*x*x;}
+  QuadraticRegression::QuadraticRegression() :
+    a_(0), b_(0), c_(0), chi_squared_(0) {}
 
-    double QuadraticRegression::getA() const {return a_;}
-    double QuadraticRegression::getB() const {return b_;}
-    double QuadraticRegression::getC() const {return c_;}
-    double QuadraticRegression::getChiSquared() const {return chi_squared_;}
+  double QuadraticRegression::eval(double x) const {return a_ + b_*x + c_*x*x;}
 
+  double QuadraticRegression::eval(double A, double B, double C, double x) {return A + B*x + C*x*x;}
 
-} //OpenMS //Math
+  double QuadraticRegression::getA() const {return a_;}
+  double QuadraticRegression::getB() const {return b_;}
+  double QuadraticRegression::getC() const {return c_;}
+  double QuadraticRegression::getChiSquared() const {return chi_squared_;}
+
+
+  void QuadraticRegression::computeRegression(std::vector<double>::const_iterator x_begin, 
+    std::vector<double>::const_iterator x_end, 
+    std::vector<double>::const_iterator y_begin)
+  {
+    std::vector<double> weights(std::distance(x_begin, x_end), 1);
+    computeRegressionWeighted(x_begin, x_end, y_begin, weights.begin());
+  }
+
+  void QuadraticRegression::computeRegressionWeighted(
+    std::vector<double>::const_iterator x_begin, 
+    std::vector<double>::const_iterator x_end, 
+    std::vector<double>::const_iterator y_begin, 
+    std::vector<double>::const_iterator w_begin)
+  {
+    // Compute the linear fit of a quadratic function.
+    // Get the coefficients for y = w_1*a +w_2*b*x + w_3*c*x^2.
+
+    std::vector<Wm5::Vector2d> points;
+    for(std::vector<double>::const_iterator xIter = x_begin, yIter = y_begin; xIter!=x_end; ++xIter, ++yIter)
+    {
+      points.push_back( Wm5::Vector2d(*xIter, *yIter) );
+    }
+
+    // Compute sums for linear system. copy&paste from GeometricTools Wm5ApprLineFit2.cpp
+    // and modified to allow quadratic functions
+    int numPoints = static_cast<Int>(points.size());
+    double sumX = 0, sumXX = 0, sumXXX = 0, sumXXXX = 0;
+    double sumY = 0, sumXY = 0, sumXXY = 0;
+    double sumW = 0;
 
+    auto wIter = w_begin;
+    for (int i = 0; i < numPoints; ++i, ++wIter)
+    {
+
+      double x = points[i].X();
+      double y = points[i].Y();
+      double weight = *wIter;
+
+      sumX += weight * x;
+      sumXX += weight * x * x;
+      sumXXX += weight * x * x * x;
+      sumXXXX += weight * x * x * x * x;
+
+      sumY += weight * y;
+      sumXY += weight * x * y;
+      sumXXY += weight * x * x * y;
```

Comment:
```suggestion
      summand = weight;
      sumY += summand * y;
      summand *= x;
      sumX += summand;
      sumXY += summand * y;
      summand *= x;
      sumXX += summand;
      sumXXY += summand * y;
      summand *= x;
      sumXXX += summand;
      summand *= x;
      sumXXXX += summand;
```

Reply:
but I don't know if worth it.

Reply:
My guess is that the compiler is very smart here.

Reply:
https://godbolt.org/z/9vnah9Gb9

it is (as expected), but only in `-O2`. So in debug mode, we're stuck with slow code.